File: lib/evva/templates/kotlin/events.kt
sealed class <%= class_name %>(
    event: <%= enums_class_name %>,
    val properties: Map<String, Any?>? = null,
    val destinations: Array<<%= destinations_class_name %>> = emptyArray(),
) {
    val name = event.key

    <%- events.each_with_index do |e, index| -%>
    <%- if e[:properties].count == 0 -%>
    data object <%= e[:class_name] %> : <%= class_name %>(
    <%- else -%>
    data class <%= e[:class_name] %>(
        <%- e[:properties].each_with_index do |p, index| -%>
        <%= "val #{p[:param_name]}: #{p[:type]}" %>,
        <%- end -%>
    ) : <%= class_name %>(
    <%- end -%>
        event = <%= enums_class_name %>.<%= e[:event_name] %>,
        <%- if e[:properties].count > 0 -%>
        properties = mapOf(
            <%- e[:properties].each_with_index do |p, index| -%>
            "<%= p[:name] %>" to <%= p[:value_fetcher] %>,
            <%- end -%>
        ),
        <%- end -%>
        <%- if e[:destinations].count > 0 -%>
        destinations = arrayOf(
            <%- e[:destinations].each_with_index do |d, index| -%>
            <%= destinations_class_name %>.<%= d %>,
            <%- end -%>
        ),
        <%- end -%>
    )
    <%- unless index == events.count - 1 -%>

    <%- end -%>
    <%- end -%>
}

File: lib/evva/templates/swift/events.swift
struct EventData {
	let name: String
	var properties: [String: Any]?
	let destinations: [Destination]

	init(name: String, properties: [String: Any]?, destinations: [Destination]) {
		self.name = name
		self.properties = properties
		self.destinations = destinations
	}

	init(type: EventType, properties: [String: Any]?) {
		self.init(name: type.name, properties: properties, destinations: type.destinations)
	}
}

enum EventType: String {
	<%- events.each do |e| -%>
	case <%= e[:case_name] %> = "<%= e[:event_name] %>"
	<%- end -%>

	var name: String { return rawValue }

	var destinations: [Destination] {
		switch self {
		<%- events.each do |e| -%>
		case .<%= e[:case_name] %>: return [<%= e[:destinations].map { |d| ".#{d}" }.join(", ") %>]
		<%- end -%>
		}
	}
}

enum Event {
	<%- events.each do |e| -%>
	<%- if e[:properties].count == 0 -%>
	case <%= e[:case_name] %>
	<%- else -%>
	case <%= e[:case_name] %>(<%= e[:properties].map { |p| "#{p[:name]}: #{p[:type]}" }.join(", ") %>)
	<%- end -%>
	<%- end -%>

	var data: EventData {
		switch self {
		<%- events.each_with_index do |e, index| -%>
		<%- if e[:properties].count == 0 -%>
		case .<%= e[:case_name] %>:
		<%- else -%>
		case let .<%= e[:case_name] %>(<%= e[:properties].map { |p| p[:name] }.join(", ") %>):
		<%- end -%>
			return EventData(type: .<%= e[:case_name] %>,
				<%- if e[:properties].count == 0 -%>
							 properties: nil)
				<%- else -%>
							 properties: [
					<%- e[:properties].each do |p| -%>
								"<%= p[:name] %>": <%= p[:value] %> as Any,
					<%- end -%>
							 ])
				<%- end -%>
				<%- unless index == events.count - 1 -%>

				<%- end -%>
		<%- end -%>
		}
	}
}
